tech/post/2017-11-06-Kubernetes-Taints&Tolerations.md

```diff
@@ -0,0 +1,302 @@
+# Taints and Tolerations
+
+## 1. Introducation
+
+Taints and tolerations work together to ensure that pods are not scheduled onto inappropriate nodes.
+One or more taints are applied to a node; this marks that the node should not accept any pods that do not tolerate the taints.
+Tolerations are applied to pods, and allow the pods to schedule onto nodes with matching taints.
+
+![](../images/2017-11-06-Kubernetes-Taints&Tolerations/Concept.png)
+
+Node affinity is a property of pods that attracts them to a set of nodes.
+Taints are the opposite and they allow a node to repel a set of pods.
+If Node affinity could be called by ```White List```, Taints could be called by ```Black List```.
+
+## 2. Operations
+
+### Basis
+
+Taints are defined for Kubernetes Cluster nodes in the ```annotations```,
+and Tolerations are specified for Kubernetes pods in the ```PodSpec```.
+
+The Scheduler will schedule the pods by lots of rules, including below:
+
+* PodToleratesNodeTaints: one of the predicates rules.
+
+  Judge whether Pod satisfy the conditions of node tolerance.
+
+* TaintTolerationPriority: one of the priorities rules.
+
+  Compare the TolerationList in Pod with the Taints in node, the more paired, the lower score.
+
+### Taints
+
+Kubernetes provides two ways to add Taints for node.
+
+* Using command to add Taints.
+
+  ```kubectl taint nodes [nodename] key=value:NoSchedule```
+
+* Declare annotations for node.
+
+  ```annotations: scheduler.alpha.kubernetes.io/taints: '[{"key": "alias", "operator": "Equal"}, "value": "edison", "effect": "NoSchedule" ]'```
+
+The users can use the following command to delete Taints from node.
+
+* Using command to delete Taints.
+
+  ```kubectl taint nodes [nodename] key:NoSchedule-```
+
+```operator``` could be defined by：
+
+* Equal
+
+  Whether the ```key``` is equal with the ```value```. This is the default ```Operator```.
+
+* Exists
+
+  Whether the ```key``` is existing. It is not necessary to specify the ```value```.
+
+```effect``` could be defined by:
+
+* NoSchedule
+
+  If the Tolerations declared in the Pods don't match the Taints defined in the Node,
+  it will not allow the Pods to schedule on the Node.

```

Comment:
allow the Pods to be scheduled on the Node

Reply:
Done. Modify it into 'be scheduled onto the Node'